packager/certcache/storage.go

```diff
@@ -71,23 +105,33 @@ func (this *LocalFile) Read(ctx context.Context, isExpired func([]byte) bool, up
 			return contents, nil
 		}
 		// Upgrade to a write-lock. It seems this may or may not be atomic, depending on the system.
+		// Windows does not handle a lock "upgrade", hence unlock before lock.
```

Comment:
This looks good to me, but let me think out loud so you can double-check my work:

I guess there's a risk of a race in the gap between the Unlock and TryLock, but as I noted on the line above, that risk probably already existed. And I think it's not a problem since it's okay to grab two OCSP responses and write them in succession. (Atomic read-modify-write isn't strictly required here; just an optimization against hammering the OCSP responder from all N instances at once. On further thought, it may be a good idea to add some randomness to the ocspCheckInterval in certcache.go, just to avoid hammering the lock file.)

No risk of deadlock, since only one lock file is in use. (I could be wrong; I often am when it comes to concurrency.)

Reply:
This is certainly my least favorite part of this commit series. As you noted, though: _that risk probably already existed_. Limiting this change to Windows means we aren't making things worse, as the attempts to transition lock type were causing `locked` to be false and so this process _would_ exit.

_No risk of deadlock, since only one lock file is in use_ : this thought occurred to me as well, since we will soon test running multiple instances of amppkg with a shared OCSP cache file. It is part of the reason I opted for hard-coding the `__.lock` filename instead of allowing it to be configurable.

All said, as much as I don't _like_ many of the changes in this commit series, I do think they introduce better compatibility without sacrificing anything on non-Windows systems.

Reply:
Yep, I'm happy that amppkg is gaining support for Windows (no judgment :). In general, I'd love to minimize complexity, but not at the cost of correctness. Thanks for confirming. And, ugh, sorry for the overly-complicated certcache code. To quote [someone](https://twitter.com/cramforce/status/949054860166553600), "I was young and needed to ship."